File: 프로그래머스/1/92334. 신고 결과 받기/신고 결과 받기.py
from collections import defaultdict
def solution(id_list, reports, k):
    answer = [0] * len(id_list)
    report_count = defaultdict(int)
    report_status = defaultdict(list)
    reports = set(reports)
    
    #k번 이상 신고 받으면 이용 정지 당함
    for report in reports:
        a, b = report.split(" ")
        report_count[b] += 1
        report_status[a].append(b)
    
    for i in range(len(id_list)):
        for man in report_status[id_list[i]]:
            if report_count[man] >= k:
                answer[i] += 1
    
    return answer

File: 프로그래머스/1/92334. 신고 결과 받기/신고 결과 받기.java
import java.util.*;
class Solution {
    public int[] solution(String[] id_list, String[] report, int k) {
        int[] answer = new int[id_list.length];
        HashMap<String, Integer> map = new HashMap<>();
        HashMap<String, ArrayList<String>> singo = new HashMap<>();
        HashSet<String> set = new HashSet<>(Arrays.asList(report));
        
        for (String r : set) {
            String[] split = r.split(" ");
            String from = split[0];
            String to = split[1];

            map.put(to, map.getOrDefault(to, 0) + 1);

            singo.putIfAbsent(from, new ArrayList<>());
            singo.get(from).add(to);
        }
        
        ArrayList<Integer> block = new ArrayList<>();
        for(String s: id_list) {
            int cnt = 0;
            if (!singo.containsKey(s)) {
                block.add(0);
            }
            else {
                for(String b : singo.get(s)) {
                    if(map.get(b) >= k) {
                        cnt++;
                    }
                }
                block.add(cnt);
            }
        }
        return block.stream().mapToInt(i->i).toArray();
    }
}